File: 2026/09/26/flutter-gpt-live-senses/media_taps/macos/media_taps/Sources/media_taps/MediaTapsPlugin.swift
// Shared by iOS and macOS (macos/ holds an identical copy).
import AVFoundation
import WebRTC
import flutter_webrtc

#if os(iOS)
  import Flutter
#else
  import FlutterMacOS
#endif

/// Attaches sinks to flutter_webrtc's native tracks. Counters (and optional
/// bounded copies) are pulled by Dart with `poll`. Channel calls run on the
/// main thread, where flutter_webrtc also mutates its track dictionaries.
public class MediaTapsPlugin: NSObject, FlutterPlugin {
  private var taps: [Int: Tap] = [:]
  private var nextId = 1

  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(iOS)
      let messenger = registrar.messenger()
    #else
      let messenger = registrar.messenger
    #endif
    let channel = FlutterMethodChannel(name: "media_taps", binaryMessenger: messenger)
    registrar.addMethodCallDelegate(MediaTapsPlugin(), channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "attach":
      do {
        result(try attach(args))
      } catch {
        result(FlutterError(code: "media_taps", message: "\(error)", details: nil))
      }
    case "detach":
      if let id = args["id"] as? Int { taps.removeValue(forKey: id)?.detach() }
      result(nil)
    case "poll":
      result(taps.keys.sorted().compactMap { taps[$0]?.drain() })
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private struct TapError: Error, CustomStringConvertible {
    let description: String
  }

  private func attach(_ args: [String: Any]) throws -> Int {
    guard let trackId = args["trackId"] as? String,
      let kind = args["kind"] as? String,
      let local = args["local"] as? Bool
    else { throw TapError(description: "bad arguments") }
    let convertWidth = args["convertWidth"] as? Int ?? 0
    let deliver = args["deliver"] as? Bool ?? false
    guard let plugin = FlutterWebRTCPlugin.sharedSingleton() else {
      throw TapError(description: "flutter_webrtc is not registered")
    }
    let localTrack = plugin.localTracks?[trackId] as? LocalTrack
    let id = nextId
    nextId += 1
    let tap: Tap
    if kind == "video" {
      let track = local ? localTrack?.track() : plugin.remoteTrack(forId: trackId)
      guard let video = track as? RTCVideoTrack else {
        throw TapError(description: "no video track \(trackId)")
      }
      tap = VideoTap(id: id, track: video, convertWidth: convertWidth, deliver: deliver)
    } else if local {
      // A renderer on a local RTCAudioTrack receives nothing (the local
      // source's AddSink is a no-op); flutter_webrtc routes local audio
      // through the audio processing module instead. This is process-wide.
      guard let audio = localTrack as? LocalAudioTrack else {
        throw TapError(description: "no local audio track \(trackId)")
      }
      tap = LocalAudioTap(id: id, track: audio, deliver: deliver)
    } else {
      guard let audio = plugin.remoteTrack(forId: trackId) as? RTCAudioTrack else {
        throw TapError(description: "no remote audio track \(trackId)")
      }
      tap = RemoteAudioTap(id: id, track: audio, deliver: deliver)
    }
    taps[id] = tap
    return id
  }
}

private class Tap: NSObject {
  let id: Int
  let kind: String
  let lock = NSLock()
  var counters: [String: Any] = [:]

  init(id: Int, kind: String) {
    self.id = id
    self.kind = kind
    super.init()
  }

  func inc(_ key: String, _ by: Int = 1) {
    counters[key] = (counters[key] as? Int ?? 0) + by
  }

  func detach() {}
  func drain() -> [String: Any] { [:] }
}

private func fourCC(_ value: OSType) -> String {
  let bytes = [24, 16, 8, 0].map { UInt8((value >> $0) & 0xff) }
  return String(bytes: bytes, encoding: .ascii) ?? "\(value)"
}

private final class VideoTap: Tap, RTCVideoRenderer {
  static let maxFrames = 8
  private let track: RTCVideoTrack
  private let convertWidth: Int
  private let deliver: Bool
  private var queue: [[String: Any]] = []

  init(id: Int, track: RTCVideoTrack, convertWidth: Int, deliver: Bool) {
    self.track = track
    self.convertWidth = convertWidth
    self.deliver = deliver
    super.init(id: id, kind: "video")
    track.add(self)
  }

  func setSize(_ size: CGSize) {}

  func renderFrame(_ frame: RTCVideoFrame?) {
    guard let frame = frame else { return }
    let buffer = frame.buffer
    let w = Int(buffer.width)
    let h = Int(buffer.height)
    let bufferType: String
    if let cv = buffer as? RTCCVPixelBuffer {
      bufferType = "cvpixelbuffer-" + fourCC(CVPixelBufferGetPixelFormatType(cv.pixelBuffer))
    } else if buffer is RTCI420Buffer {
      bufferType = "i420"
    } else {
      bufferType = String(describing: type(of: buffer))
    }
    var luma: Data? = nil
    var dw = 0
    var dh = 0
    var us = 0
    if convertWidth != 0 {
      let t0 = DispatchTime.now().uptimeNanoseconds
      dw = convertWidth < 0 ? w : convertWidth
      dh = (h * dw / w) & ~1
      let scaled: RTCVideoFrameBuffer
      if dw != w || dh != h,
        let s = buffer.cropAndScale?(
          with: 0, offsetY: 0, cropWidth: Int32(w), cropHeight: Int32(h),
          scaleWidth: Int32(dw), scaleHeight: Int32(dh))
      {
        scaled = s
      } else {
        scaled = buffer
        dw = w
        dh = h
      }
      let i420 = scaled.toI420()
      var out = Data(count: dw * dh)
      out.withUnsafeMutableBytes { raw in
        let dst = raw.bindMemory(to: UInt8.self).baseAddress!
        let stride = Int(i420.strideY)
        for row in 0..<dh {
          (dst + row * dw).update(from: i420.dataY + row * stride, count: dw)
        }
      }
      luma = out
      us = Int((DispatchTime.now().uptimeNanoseconds - t0) / 1000)
    }
    lock.lock()
    defer { lock.unlock() }
    inc("callbacks")
    counters["width"] = w
    counters["height"] = h
    counters["rotation"] = frame.rotation.rawValue
    counters["bufferType"] = bufferType
    if convertWidth != 0 {
      inc("convertCount")
      inc("convertUsTotal", us)
      counters["convertUsMax"] = max(counters["convertUsMax"] as? Int ?? 0, us)
    }
    if deliver, let luma = luma {
      if queue.count >= Self.maxFrames {
        queue.removeFirst()
        inc("droppedToDart")
      }
      queue.append(["width": dw, "height": dh, "luma": FlutterStandardTypedData(bytes: luma)])
    }
  }

  override func detach() { track.remove(self) }

  override func drain() -> [String: Any] {
    lock.lock()
    defer { lock.unlock() }
    let frames = queue
    queue.removeAll()
    return ["id": id, "kind": kind, "counters": counters, "frames": frames]
  }
}

private class AudioTap: Tap {
  static let maxPcmBytes = 96_000
  private let deliver: Bool
  private var sumSquares = 0.0
  private var samples = 0
  private var pcm = Data()

  init(id: Int, deliver: Bool) {
    self.deliver = deliver
    super.init(id: id, kind: "audio")
  }

  /// Called with interleaved Int16 samples.
  func record(
    _ data: UnsafePointer<Int16>, frames: Int, channels: Int, sampleRate: Int
  ) {
    let n = frames * channels
    var sq = 0.0
    for i in 0..<n {
      let s = Double(data[i])
      sq += s * s
    }
    lock.lock()
    defer { lock.unlock() }
    inc("callbacks")
    counters["sampleRate"] = sampleRate
    counters["channels"] = channels
    counters["framesPerChunk"] = frames
    counters["bitsPerSample"] = 16
    sumSquares += sq
    samples += n
    if deliver {
      if pcm.count + n * 2 > Self.maxPcmBytes {
        inc("droppedToDart")
      } else {
        data.withMemoryRebound(to: UInt8.self, capacity: n * 2) {
          pcm.append($0, count: n * 2)
        }
      }
    }
  }

  override func drain() -> [String: Any] {
    lock.lock()
    defer { lock.unlock() }
    counters["rms"] = samples == 0 ? NSNull() : (sqrt(sumSquares / Double(samples)) / 32768.0) as Any
    sumSquares = 0
    samples = 0
    var out: [String: Any] = ["id": id, "kind": kind, "counters": counters]
    if !pcm.isEmpty {
      out["pcm"] = FlutterStandardTypedData(bytes: pcm)
      pcm = Data()
    }
    return out
  }
}

/// Mic after echo cancellation / noise suppression, via flutter_webrtc's
/// capture post-processing hook (process-wide, not per track).
private final class LocalAudioTap: AudioTap, ExternalAudioProcessingDelegate {
  private let track: LocalAudioTrack
  private var sampleRate = 0
  private var scratch: [Int16] = []

  init(id: Int, track: LocalAudioTrack, deliver: Bool) {
    self.track = track
    super.init(id: id, deliver: deliver)
    track.addProcessing(self)
  }

  func audioProcessingInitialize(withSampleRate sampleRateHz: Int, channels: Int) {
    sampleRate = sampleRateHz
  }

  func audioProcessingProcess(_ audioBuffer: RTCAudioBuffer) {
    // Channel 0 only; float samples on the int16 scale.
    let frames = Int(audioBuffer.frames)
    let src = audioBuffer.rawBuffer(forChannel: 0)
    if scratch.count < frames { scratch = [Int16](repeating: 0, count: frames) }
    for i in 0..<frames {
      scratch[i] = Int16(max(-32768, min(32767, src[i])))
    }
    let rate = sampleRate > 0 ? sampleRate : frames * 100
    scratch.withUnsafeBufferPointer {
      record($0.baseAddress!, frames: frames, channels: 1, sampleRate: rate)
    }
  }

  func audioProcessingRelease() {}

  override func detach() { track.removeProcessing(self) }
}

/// Decoded per-track PCM before mixing.
private final class RemoteAudioTap: AudioTap, RTCAudioRenderer {
  private let track: RTCAudioTrack

  init(id: Int, track: RTCAudioTrack, deliver: Bool) {
    self.track = track
    super.init(id: id, deliver: deliver)
    track.add(self)
  }

  func render(pcmBuffer: AVAudioPCMBuffer) {
    guard let data = pcmBuffer.int16ChannelData else { return }
    // RTCAudioRendererAdapter delivers interleaved Int16 in channel 0.
    record(
      data[0], frames: Int(pcmBuffer.frameLength),
      channels: Int(pcmBuffer.format.channelCount),
      sampleRate: Int(pcmBuffer.format.sampleRate))
  }

  override func detach() { track.remove(self) }
}

File: 2026/09/25/flutter-webrtc-media-taps/media_taps/android/src/main/kotlin/com/kiarina/labs/media_taps/MediaTapsPlugin.kt
package com.kiarina.labs.media_taps

import com.cloudwebrtc.webrtc.FlutterWebRTCPlugin
import com.cloudwebrtc.webrtc.audio.LocalAudioTrack
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.webrtc.AudioTrack
import org.webrtc.AudioTrackSink
import org.webrtc.VideoFrame
import org.webrtc.VideoSink
import org.webrtc.VideoTrack
import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Attaches sinks to flutter_webrtc's native tracks. Everything the sinks see is
 * summarised in counters (and optionally copied into bounded queues); Dart
 * pulls it with `poll`. All channel calls run on the platform thread, where
 * flutter_webrtc also mutates its track maps.
 */
class MediaTapsPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private val taps = HashMap<Int, Tap>()
    private var nextId = 1

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "media_taps")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        taps.values.forEach { it.detach() }
        taps.clear()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "attach" -> result.success(attach(call))
                "detach" -> {
                    taps.remove(call.argument<Int>("id"))?.detach()
                    result.success(null)
                }
                "poll" -> result.success(taps.values.map { it.drain() })
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("media_taps", e.toString(), null)
        }
    }

    private fun attach(call: MethodCall): Int {
        val trackId = call.argument<String>("trackId")!!
        val kind = call.argument<String>("kind")!!
        val local = call.argument<Boolean>("local")!!
        val convertWidth = call.argument<Int>("convertWidth") ?: 0
        val deliver = call.argument<Boolean>("deliver") ?: false
        val plugin = FlutterWebRTCPlugin.sharedSingleton
            ?: throw IllegalStateException("flutter_webrtc is not registered")
        val id = nextId++
        // Remote tracks: use the stream-backed registry. Wrappers from the
        // transceiver fallback are disposed by later getTransceivers() calls,
        // which silently removes their sinks.
        val tap: Tap = if (kind == "video") {
            val track = (if (local) plugin.getLocalTrack(trackId)?.track else plugin.getRemoteTrack(trackId))
                as? VideoTrack ?: throw IllegalArgumentException("no video track $trackId")
            VideoTap(id, track, convertWidth, deliver)
        } else if (local) {
            val track = plugin.getLocalTrack(trackId) as? LocalAudioTrack
                ?: throw IllegalArgumentException("no local audio track $trackId")
            LocalAudioTap(id, track, deliver)
        } else {
            val track = plugin.getRemoteTrack(trackId) as? AudioTrack
                ?: throw IllegalArgumentException("no remote audio track $trackId")
            RemoteAudioTap(id, track, deliver)
        }
        taps[id] = tap
        return id
    }
}

private abstract class Tap(val id: Int, val kind: String) {
    protected val lock = Any()
    protected val counters = HashMap<String, Any?>()

    protected fun inc(key: String, by: Long = 1) {
        counters[key] = ((counters[key] as? Long) ?: 0L) + by
    }

    abstract fun detach()
    abstract fun drain(): Map<String, Any?>
}

private class VideoTap(
    id: Int,
    private val track: VideoTrack,
    private val convertWidth: Int,
    private val deliver: Boolean,
) : Tap(id, "video"), VideoSink {
    private val queue = ArrayDeque<Map<String, Any>>()

    init {
        track.addSink(this)
    }

    override fun onFrame(frame: VideoFrame) {
        val buffer = frame.buffer
        val w = buffer.width
        val h = buffer.height
        var luma: ByteArray? = null
        var dw = 0
        var dh = 0
        var us = 0L
        if (convertWidth != 0) {
            val t0 = System.nanoTime()
            dw = if (convertWidth < 0) w else convertWidth
            dh = (h * dw / w) and 1.inv()
            // cropAndScale first: on a TextureBuffer it is only a matrix change,
            // so the GPU readback in toI420() happens at the small size.
            val scaled = if (dw == w && dh == h) buffer.also { it.retain() }
            else buffer.cropAndScale(0, 0, w, h, dw, dh)
            val i420 = scaled.toI420()
            scaled.release()
            if (i420 != null) {
                val out = ByteArray(dw * dh)
                val y = i420.dataY
                val stride = i420.strideY
                for (row in 0 until dh) {
                    y.position(row * stride)
                    y.get(out, row * dw, dw)
                }
                i420.release()
                luma = out
            }
            us = (System.nanoTime() - t0) / 1000
        }
        synchronized(lock) {
            inc("callbacks")
            counters["width"] = w
            counters["height"] = h
            counters["rotation"] = frame.rotation
            counters["bufferType"] = when (buffer) {
                is VideoFrame.TextureBuffer -> "texture-${buffer.type.name.lowercase()}"
                is VideoFrame.I420Buffer -> "i420"
                else -> buffer.javaClass.simpleName
            }
            if (convertWidth != 0) {
                inc("convertCount")
                inc("convertUsTotal", us)
                counters["convertUsMax"] = max((counters["convertUsMax"] as? Long) ?: 0L, us)
            }
            if (deliver && luma != null) {
                if (queue.size >= MAX_FRAMES) {
                    queue.removeFirst()
                    inc("droppedToDart")
                }
                queue.addLast(mapOf("width" to dw, "height" to dh, "luma" to luma))
            }
        }
    }

    override fun detach() {
        if (!track.isDisposed) track.removeSink(this)
    }

    override fun drain(): Map<String, Any?> = synchronized(lock) {
        val frames = queue.toList()
        queue.clear()
        mapOf("id" to id, "kind" to kind, "counters" to HashMap(counters), "frames" to frames)
    }

    companion object {
        const val MAX_FRAMES = 8
    }
}

private abstract class AudioTap(id: Int, private val deliver: Boolean) : Tap(id, "audio"), AudioTrackSink {
    private var sumSquares = 0.0
    private var samples = 0L
    private val pcm = java.io.ByteArrayOutputStream()

    override fun onData(
        audioData: ByteBuffer, bitsPerSample: Int, sampleRate: Int,
        numberOfChannels: Int, numberOfFrames: Int, absoluteCaptureTimestampMs: Long,
    ) {
        val n = numberOfFrames * numberOfChannels
        val bytes = n * bitsPerSample / 8
        val data = audioData.duplicate().order(ByteOrder.LITTLE_ENDIAN)
        var sq = 0.0
        if (bitsPerSample == 16) {
            for (i in 0 until n) {
                val s = data.getShort(data.position() + i * 2).toDouble()
                sq += s * s
            }
        }
        synchronized(lock) {
            inc("callbacks")
            counters["sampleRate"] = sampleRate
            counters["channels"] = numberOfChannels
            counters["framesPerChunk"] = numberOfFrames
            counters["bitsPerSample"] = bitsPerSample
            sumSquares += sq
            samples += n
            if (deliver) {
                if (pcm.size() + bytes > MAX_PCM_BYTES) {
                    inc("droppedToDart")
                } else {
                    val chunk = ByteArray(bytes)
                    data.position(data.position())
                    data.get(chunk, 0, minOf(bytes, data.remaining()))
                    pcm.write(chunk)
                }
            }
        }
    }

    override fun drain(): Map<String, Any?> = synchronized(lock) {
        counters["rms"] = if (samples == 0L) null else sqrt(sumSquares / samples) / 32768.0
        sumSquares = 0.0
        samples = 0
        val out = mutableMapOf<String, Any?>("id" to id, "kind" to kind, "counters" to HashMap(counters))
        if (pcm.size() > 0) {
            out["pcm"] = pcm.toByteArray()
            pcm.reset()
        }
        out
    }

    companion object {
        const val MAX_PCM_BYTES = 96_000
    }
}

/** Raw mic samples fanned out by flutter_webrtc's LocalAudioTrack (before APM). */
private class LocalAudioTap(id: Int, private val track: LocalAudioTrack, deliver: Boolean) :
    AudioTap(id, deliver) {
    init {
        track.addSink(this)
    }

    override fun detach() = track.removeSink(this)
}

/** Decoded per-track PCM before mixing (native AudioTrack sink). */
private class RemoteAudioTap(id: Int, private val track: AudioTrack, deliver: Boolean) :
    AudioTap(id, deliver) {
    init {
        track.addSink(this)
    }

    override fun detach() {
        if (!track.isDisposed) track.removeSink(this)
    }
}
